src/openms/source/FORMAT/MzQCFile.cpp

```diff
@@ -0,0 +1,199 @@
+    out["MzQC"]["runQualities"]["metadata"]["analysisSoftware"] = { // todo
+        {
+          {"accession", "MS:1009001" }, // create new qc-cv for QCCalculator: MS:1009001 quality control metrics generating software
+          {"name", "QCCalculator"},
+          {"version", String(version.version_major)+"."+String(version.version_minor)+"."+String(version.version_patch)},
+          {"uri", "https://www.openms.de"}
+        }
+    };
+
+    out["MzQC"]["runQualities"]["qualityMetrics"] = {};
+    
+    out["MzQC"]["controlledVocabularies"] = {
+        {
+          {"name", "Proteomics Standards Initiative Quality Control Ontology"},
+          {"uri", "https://github.com/HUPO-PSI/mzQC/blob/master/cv/qc-cv.obo"},
+          {"version", "0.1.2"},
+        },
+        {
+          {"name", "Proteomics Standards Initiative Mass Spectrometry Ontology"},
+          {"uri", "https://github.com/HUPO-PSI/psi-ms-CV/blob/master/psi-ms.obo"},
+          {"version", "4.1.49"}
+        }
+    };
+
+    // precursors QC:0000044 not in cv any more
+
+    map<Size, UInt> counts;
+    for (const auto& spectrum : exp)
+      {
+        const Size level = spectrum.getMSLevel();
+        ++counts[level];  // count MS level
+      }
+
+    // create qualityMetric in json format
+    auto addValue = [](const String& accession, const String& name, const auto& value) -> json
+      {
+        ControlledVocabulary cv;
+        cv.loadFromOBO("PSI-MS", File::find("/CV/psi-ms.obo"));
+        cv.loadFromOBO("QC", File::find("/CV/qc-cv.obo"));
+        json qm;
+        qm["accession"] = accession;
+        if (cv.exists(accession)) 
+        {
+          qm["name"] = cv.getTerm(accession).name;
+        } else
+        {
+          qm["name"] = name;
+          cout << accession << " not found in CV." << endl;
+        }
+        qm["value"] = value;
+        return qm;
+      };
+
+    // Number of MS1 spectra
+    out["MzQC"]["runQualities"]["qualityMetrics"] += addValue("QC:4000059", "Number of MS1 spectra", counts[1]);
+    // Number of MS2 spectra
+    out["MzQC"]["runQualities"]["qualityMetrics"] += addValue("QC:4000060", "Number of MS2 spectra", counts[2]);
+    // Number of chromatograms"
+    out["MzQC"]["runQualities"]["qualityMetrics"] += addValue("QC:4000135", "Number of chromatograms", exp.getChromatograms().size());
+    // Run time (RT duration)
+    out["MzQC"]["runQualities"]["qualityMetrics"] += addValue("QC:4000053", "RT duration", UInt(exp.getMaxRT() - exp.getMinRT()));
+    // MZ acquisition range
+    out["MzQC"]["runQualities"]["qualityMetrics"] += addValue("QC:4000138", "MZ acquisition range", tuple<UInt,UInt>{exp.getMinMZ(), exp.getMaxMZ()});
+    
+    //open stream
+    ofstream os(outputFileName.c_str());
+    if (!os)
+    {
+      throw Exception::UnableToCreateFile(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, outputFileName);
+    }
+
+    os.precision(writtenDigits<double>(0.0));
```

Comment:
is this correct?

Reply:
copied that from the qcML code. perhaps this is not even necessary since the conversion to a string is handled by the json library?